File: ts-sdk/src/cache/get.json.rpc.cache.provider.ts
import { FetchRequest, JsonRpcProvider, Network } from "ethers";
import { assertIsNonEmptyString } from "../utils";

export type GetJsonRpcCacheProvider = (chainId: number) => JsonRpcProvider;

export function provideGetJsonRpcCacheProvider(
  jsonRpcCacheUrl: string
): GetJsonRpcCacheProvider {
  assertIsNonEmptyString(jsonRpcCacheUrl, "jsonRpcCacheUrl");

  // store the created providers in-memory for reuse
  const jsonRpcCacheProviders: { [chainId: number]: JsonRpcProvider } = {};

  return function getJsonRpcCacheProvider(chainId: number) {
    // check if we've already created a provider for this chain
    if (chainId in jsonRpcCacheProviders) {
      return jsonRpcCacheProviders[chainId];
    }

    // create a new provider for the chain
    const rpcConnection = new FetchRequest(jsonRpcCacheUrl);
    // set the header to tell the scan node which chain we are talking about
    rpcConnection.setHeader("X-Forta-Chain-ID", chainId);
    const provider = new JsonRpcProvider(rpcConnection, Network.from(chainId), {
      staticNetwork: true,
    });
    jsonRpcCacheProviders[chainId] = provider;
    return provider;
  };
}

File: py-sdk/src/forta_bot_sdk/cache/get_json_rpc_cache_provider.py
from typing import Callable
from web3 import AsyncWeb3
from web3.middleware import async_geth_poa_middleware
from ..utils import assert_exists, assert_is_non_empty_string, RetryOptions


GetJsonRpcCacheProvider = Callable[[int], AsyncWeb3]


def provide_get_json_rpc_cache_provider(json_rpc_cache_url: str, json_rpc_cache_retry_options: RetryOptions) -> GetJsonRpcCacheProvider:
    assert_is_non_empty_string(json_rpc_cache_url, 'json_rpc_cache_url')
    assert_exists(json_rpc_cache_retry_options, 'json_rpc_cache_retry_options')

    # maintain reference to created providers
    providers: dict[int, AsyncWeb3] = {}

    def get_json_rpc_cache_provider(chain_id: int) -> AsyncWeb3:
        nonlocal providers
        # check if we've already created a provider for this chain
        if chain_id in providers:
            return providers[chain_id]

        # create a new provider for the chain
        provider = AsyncWeb3(AsyncWeb3.AsyncHTTPProvider(
            json_rpc_cache_url,
            request_kwargs={
                'headers': {
                    'X-Forta-Chain-ID': str(chain_id)
                },
                'timeout': json_rpc_cache_retry_options['timeout_seconds']
            }))
        providers[chain_id] = provider
        return provider

    return get_json_rpc_cache_provider
